// server/controllers/orderController.js
const Order = require('../models/Order');
const Book = require('../models/Book');
const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY);
const crypto = require('crypto');
const nodemailer = require('nodemailer');

exports.createOrder = async (req, res) => {
  const { bookId, customerName, customerEmail } = req.body;
  try {
    const book = await Book.findById(bookId);
    if (!book) return res.status(404).json({ message: 'Book not found' });

    // Mock payment successful logic if stripe secret is missing
    const paymentIntent = process.env.STRIPE_SECRET_KEY 
      ? await stripe.paymentIntents.create({
          amount: Math.round(book.price * 100),
          currency: 'usd',
          metadata: { bookId, customerEmail }
        })
      : { id: 'mock_pi_' + Date.now(), client_secret: 'mock_cs_' + Date.now() };

    const downloadToken = crypto.randomBytes(32).toString('hex');
    const order = new Order({
      customerName,
      customerEmail,
      bookId,
      amount: book.price,
      paymentIntentId: paymentIntent.id,
      downloadToken,
      downloadExpires: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000) // 7 days
    });

    await order.save();
    res.json({ clientSecret: paymentIntent.client_secret, orderId: order._id });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

exports.completeOrder = async (req, res) => {
  try {
    const { orderId } = req.body;
    const order = await Order.findById(orderId).populate('bookId');
    if (!order) return res.status(404).json({ message: 'Order not found' });

    order.status = 'completed';
    await order.save();

    // Send email logic
    const transporter = nodemailer.createTransport({
      host: process.env.EMAIL_HOST || 'smtp.ethereal.email',
      port: 587,
      auth: {
        user: process.env.EMAIL_USER,
        pass: process.env.EMAIL_PASS,
      },
    });

    const downloadUrl = `${process.env.CLIENT_URL || 'http://localhost:5173'}/api/orders/download/${order.downloadToken}`;

    await transporter.sendMail({
      from: '"ATHAR Store" <noreply@athar.ai>',
      to: order.customerEmail,
      subject: `Your Digital Purchase: ${order.bookId.title}`,
      text: `Hello ${order.customerName},\n\nThank you for your purchase. You can download your book here: ${downloadUrl}\n\nThis link expires in 7 days.\n\nBecause knowledge leaves an impact,\nATHAR Team`,
      html: `
        <div style="font-family: sans-serif; max-width: 600px; margin: auto; border: 1px solid #eee; padding: 20px;">
          <h2 style="color: #D4AF37;">ATHAR (أثر)</h2>
          <p>Hello <strong>${order.customerName}</strong>,</p>
          <p>Thank you for choosing ATHAR. Your digital book <strong>"${order.bookId.title}"</strong> is ready for download.</p>
          <div style="text-align: center; margin: 30px 0;">
            <a href="${downloadUrl}" style="background: #000; color: #fff; padding: 15px 30px; text-decoration: none; border-radius: 10px; font-weight: bold;">Download Now</a>
          </div>
          <p style="color: #888; font-size: 12px;">This link will expire in 7 days. If you have any issues, reply to this email.</p>
          <hr style="border: none; border-top: 1px solid #eee; margin: 20px 0;" />
          <p style="text-align: center; color: #D4AF37; font-style: italic;">"Because knowledge leaves an impact"</p>
        </div>
      `,
    });

    res.json({ message: 'Order completed', downloadLink: `/api/orders/download/${order.downloadToken}` });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

exports.downloadBook = async (req, res) => {
  try {
    const { token } = req.params;
    const order = await Order.findOne({ downloadToken: token, status: 'completed' }).populate('bookId');
    
    if (!order) return res.status(404).json({ message: 'Invalid or expired download link' });
    if (new Date() > order.downloadExpires) return res.status(403).json({ message: 'Link expired' });

    const fs = require('fs');
    const path = require('path');
    const filePath = path.join(__dirname, '..', order.bookId.pdfFile);
    
    if (!fs.existsSync(filePath)) return res.status(404).json({ message: 'File not found on server' });

    res.download(filePath, `${order.bookId.title}.pdf`);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};
